### ISSUE TITLE
Add child chain validation to the client

### ISSUE THREAD

**Issue by DavidKnott** [omisego/plasma-mvp#32]
The command line interface is able to hit the child chain server and receive blocks up until the most recent block, the next step is to add validation on the client side to make sure that each block being received from the child chain is valid (because they're not everyone should exit).

**Comment by nvonpentz** [comment]
Thanks for all your work on this. I have a few thoughts.

The client will do many of the same transaction validations that the child chain server does. Instead of duplicating the validation logic in the client, is it favorable to modularize the validation code so it can be used in both?

Related, it looks like the child chain server doesn't have a database, and is storing the blockchain in memory.  Is this intended?

https://github.com/omisego/plasma-mvp/blob/569e880178d20e589b35161e305c7e698e48129b/plasma/child_chain/child_chain.py#L17

This is different from the client which stores blocks in a database.

https://github.com/omisego/plasma-mvp/blob/569e880178d20e589b35161e305c7e698e48129b/plasma/cli/client.py#L50

Keeping this consistent may make it easier to reuse the code, if that's the direction you want to go.

**Comment by kfichter** [comment]
@nvonpentz @DavidKnott I definitely agree that the server should have a database. Whether or not the CLI should store state is dependent on what we want the CLI to do. My opinion is that the CLI should also be stateless and simply make calls to the server via the client wrapper.

Some of the design decisions behind the first iterations of Plasma will definitely effect what the client looks like. If we assume that the operator is centralized (for now), then our architecture should look something like this:

- Operator server. Effectively the current iteration of `child_chain`.
- Plasma daemons that constantly pull blocks from the operator & expose an RPC interface for clients. This basically involves running `child_chain` alongside a daemon that receives blocks from the operator and submits them to the `child_chain`.
-  Clients. Don't store any state, just rely on the local `child_chain` to store state for them.

**Comment by nvonpentz** [comment]
@kfichter Thanks for the quick response. I like your approach; having the users and the operator run different instances of the same child chain that each individually store state makes sense to me.

Do you want to have the client keep track of a local child chain in addition to the child chain service that interacts with the operator's child chain:

https://github.com/omisego/plasma-mvp/blob/569e880178d20e589b35161e305c7e698e48129b/plasma/client/client.py#L17

This local child chain would store state as you described and the CLI wouldn't need to store state as it does currently:

https://github.com/omisego/plasma-mvp/blob/569e880178d20e589b35161e305c7e698e48129b/plasma/cli/client.py#L50

When the operator submits a new block to the root chain, the user should get the corresponding block from the operator's plasma chain (do you want to add a `SubmitBlock` event?).  The user then checks if each of the transactions are valid by cross referencing it's own local child chain, and making sure the block hash matches the block hash submitted to the root chain.  If there is no problems, add the block to the user's local child chain.  Otherwise, it automatically initiates an exit.

Is that roughly what you had in mind?

**Comment by kfichter** [comment]
@nvonpentz 

> The user then checks if each of the transactions are valid by cross referencing it's own local child chain, and making sure the block hash matches the block hash submitted to the root chain. If there is no problems, add the block to the user's local child chain. Otherwise, it automatically initiates an exit.

Exactly. In the long term, the idea should be that the network is completely p2p. Operators broadcast blocks whenever they're submitted to the root chain. Users receive or request those blocks and double check them against their own local chain. Users will exit if something goes wrong (users see a hash published to the root chain but don't receive the block within some amount of time, block is invalid, etc.). 

In the short term, this probably won't be a p2p network but a direct connection to the operator. That definitely opens the operator up to some DoS attacks, but that's fine for now.
